pbft/pbft.cpp

```diff
@@ -978,6 +992,533 @@ pbft::get_high_water_mark()
+
+
+    if (ours.pre_prepare_messages_size() != thiers.pre_prepare_messages_size() )
+    {
+        LOG(error) << "is_valid_newview_message - expected " << ours.pre_prepare_messages_size() << " preprepares in new view, found " << thiers.pre_prepare_messages_size();
+        return false;
+    }
+
+    for (int i{0};i < thiers.pre_prepare_messages_size();++i)
+    {
+        if (!this->crypto->verify(thiers.pre_prepare_messages(i)))
+        {
+            LOG(error) <<  "is_valid_newview_message - unable to verify thier pre prepare message";
+            return false;
+        }
+
+        if ( thiers.pre_prepare_messages(i).sender() != original_thiers.sender())
+        {
+            LOG(error) << "is_valid_newview_message - pre prepare messaged does not come from the correct sender";
+            return false;
+        }
+
+        pbft_msg ours_pbft;
+        pbft_msg thiers_pbft;
+
+        ours_pbft.ParseFromString(ours.pre_prepare_messages(i).pbft());
+        thiers_pbft.ParseFromString(thiers.pre_prepare_messages(i).pbft());
+
+        if (ours_pbft.type() != thiers_pbft.type() || ours_pbft.view() != thiers_pbft.view() || ours_pbft.sequence() != thiers_pbft.sequence() || ours_pbft.request_hash() != thiers_pbft.request_hash())
+        {
+            LOG(error) << "is_valid_newview_message - type, view, sequence or request hash mismatch between our view change and their viewchange";
+            return false;
+        }
+    }
+    return true;
+}
+
+void
+pbft::fill_in_missing_pre_prepares(uint64_t max_checkpoint_sequence, uint64_t new_view, std::map<uint64_t, bzn_envelope> &pre_prepares) const
+{
+    uint64_t last_sequence_number{0};
+    for(const auto& pre_prepare : pre_prepares)
+    {
+        last_sequence_number = std::max(last_sequence_number, pre_prepare.first);
+    }
+
+    for (uint64_t i = max_checkpoint_sequence + 1; i < last_sequence_number; ++i)
+    {
+        //  -- create a new preprepare for a no-op operation using this sequence number
+        if (pre_prepares.find(i) == pre_prepares.end())
+        {
+            // TODO: Isabel - please review this block
+            database_msg msg;
+            msg.set_allocated_nullmsg(new database_nullmsg);
+
+            auto request = new bzn_envelope;
+            request->set_database_msg(msg.SerializeAsString());
+            this->crypto->sign(*request);
+
+            pbft_msg msg2;
+            msg2.set_view(new_view);
+            msg2.set_sequence(i);
+            msg2.set_type(PBFT_MSG_PREPREPARE);
+            msg2.set_allocated_request(request);
+            msg2.set_request_hash(this->crypto->hash(*request));
+
+            pre_prepares[i] = this->wrap_message((pbft_msg)msg2);
+        }
+    }
+}
+
+/**
+ * Create a <NEWVIEW, v+1, V, O> message
+ * @param new_view_index v + 1
+ * @param view_change_messages V
+ * @param preprepares O
+ * @return pbft_msg containing a new view message
+ */
```

Comment:
parameters in comment don't match actual parameters

Reply:
Removed the comment.